src/app/pushclient.js

```diff
@@ -1,86 +1,93 @@
+var promiseRetry = require('promise-retry');
+var Q = require('q');
+
 
 module.exports = class PushClient {
 
-  constructor(pushConfig, retryOptions) {
-    if(!pushConfig.organizationId) throw new Error(`PushClient needs an organizationId in the pushConfig.`);
-    if(!pushConfig.sourceId) throw new Error(`PushClient needs an sourceId in the pushConfig.`);
-    if(!pushConfig.pushtoken) throw new Error(`PushClient needs an pushtoken in the pushConfig.`);
-    this.organizationId = pushConfig.organizationId;
-    this.sourceId = pushConfig.sourceId;
-    this.pushtoken = pushConfig.pushtoken;
-    this.SERVER = pushConfig.server || 'push.cloud.coveo.com';
-    this.APIVERSION = pushConfig.apiversion || 'v1';
+  constructor(config, retryOptions) {
+    config = config || {}; retryOptions = retryOptions || {};
+    assert(config.organizationId, `PushClient needs an organizationId in the config.`);
+    assert(config.sourceId, `PushClient needs an sourceId in the config.`);
+    assert(config.pushApiKey, `PushClient needs an pushApiKey in the config.`);
+
+    this.organizationId = config.organizationId;
+    this.sourceId = config.sourceId;
+    this.pushApiKey = config.pushApiKey;
+    this.SERVER = config.server || 'push.cloud.coveo.com';
+    this.APIVERSION = config.apiversion || 'v1';
 
     this.RETRY = retryOptions.retry || true;
     this.RETRY_ATTEMPTS = retryOptions.attempts || 5;
     this.RETRY_DELAY = retryOptions.delay || 3000; //milliseconds
 
   }
 
-  pushDocument(document, callback) {
-    if (callback) return this.pushDocumentCallback(document, callback, 0);
+  pushDocumentWithRetry(document, callback) {
+    return (callback instanceof Function) ? this.pushDocumentCallback(document, callback) : this.pushDocumentPromise(document);
+  }
 
-    return this.pushDocumentPromise(document, 0);
+  pushDocumentCallback(document, callback) {
+    this.pushDocumentPromise(document)
+      .then(res => {
+        callback(res);
+      })
+      .catch(err => {
+        callback(undefined, err);
+      })
   }
 
-  pushDocumentCallback(document, callback, retryCount) {
-    console.log(document);
-    request({
-      url: `https://${this.SERVER}/${this.APIVERSION}/organizations/${this.organizationId}/sources/${this.sourceId}/documents`,
-      qs: { documentId: document.documentId },
-      method: 'PUT',
-      headers: {
-        'content-type': 'application/json',
-        'Authorization': 'Bearer ' + this.pushtoken
-      },
-      body: JSON.stringify(document)
-    }, (error, response, body) => {
-      if(error)
-      if (this.RETRY && retryCount < this.RETRY_ATTEMPTS) {
-        return setTimeout(() => {this.pushDocumentCallback(document, callback, ++retryCount)}, this.RETRY_DELAY);
-      } else {
-        callback(error, response, body)
-      }
-    });
+  pushDocumentPromise(document) {
+    let requestOptions = this.buildPushOptions(document);
+
+    return promiseRetry((retry, retryNumber) => {
+      return Q()
```

Comment:
Not sure why you are using Q and not Promise here, since we deal with Node.js and not browsers. But this works, we can discuss later, when I'm back.